tdct/deconvolution.py

```diff
@@ -0,0 +1,864 @@
+        if debug is True: print(clrmsg.DEBUG, "Loading images: ", datapath," , ",psfdatapath)
+        progr0.increment() #1
+
+        data_np = tf.imread(datapath)
+        psf_np = tf.imread(psfdatapath)
+
+        progr0.increment() #2
+
+        #Normalise both data (and also converts)
+        
+        data_np_norm = _convertAndNormalise(data_np)
+        psf_np_norm = _convertAndNormalise(psf_np)
+
+        progr0.increment() #3
+
+        #Determine best shape to do the fft
+        #based on fftconvolve() code at https://github.com/scipy/scipy/blob/master/scipy/signal/signaltools.py
+        #Simply use the maximum in each dimension
+        s1 = data_np_norm.shape
+        s2 = psf_np_norm.shape
+        bestshape = [max((s1[i], s2[i])) for i in range(data_np_norm.ndim)]
+        fslice = tuple([slice(sz) for sz in bestshape])
+        #It may be pssoble to accelerate the FFT by choosing higher shape values
+        #Similar to what is used in scipy _freq_domain_conv()  next_fast_len()
+        #but we are not using this trick here
+
+        psf_fft = scipy.fft.rfftn(psf_np_norm, bestshape)
+        xn1 = np.array(data_np_norm) #initialize
+
+        progr0.increment() #4
+
+        for i in range(niter):
+            xn=xn1
+
+            #convolution xn (*) psf
+            xn_fft = scipy.fft.rfftn(xn, bestshape)
+            conv1_fft = np.multiply(xn_fft , psf_fft)
+            #Hx = scipy.fft.ifftn(conv1_fft,bestshape)
+            Hx = scipy.fft.irfftn(conv1_fft,bestshape)
+            #Fix size of the result so that convolution result has the same shape as data
+            #This is needed otherwise it will fail to divide
+            Hx = _centered(Hx, data_np_norm.shape)
+
+            yhx = np.divide(data_np_norm,Hx) #Problem, Hx shape has one index higher in zz
+            
+            progr0.increment()
+
+            #correlation of the result with psf (note that is not a convolution)
+            yhx_fft = scipy.fft.rfftn(yhx, bestshape)
+            htyhx_fft = np.multiply(yhx_fft , np.conjugate(psf_fft))
+            #htyhx = scipy.fft.ifftn(htyhx_fft,bestshape)
+            htyhx = scipy.fft.irfftn(htyhx_fft,bestshape)
+            htyhx = _centered(htyhx, data_np_norm.shape)
+
+            xn1 = np.multiply(xn, htyhx)
+            
+            progr0.increment()
+
+        #At the end xn1 should have the deconvoluted data
+        
+        #Convert to uint8
+        data_deconv_norm_256 = _convertAndNormalise(xn1)*256
+        data_deconv_uint8 = data_deconv_norm_256.astype('uint8')
+        
+        progr0.increment() #5
+
+        #Save data
+        fpath,fname = os.path.split(datapath)
+        fname0 = os.path.join(fpath, "DeconvRL"+str(niter)+"it_"+fname)
+        tf.imsave(fname0 ,data_deconv_uint8)
+
+        progr0.setmax()
+ 
+        if debug is True: print(clrmsg.DEBUG, "Completed deconvolution, file saved: ", fname0)
+
+def doRLDeconvolution3(datapath , psfdatapath , niter=0, qtprocessbar=None):
```

Comment:
Ok, this is a third function. I just noticed in your email that the current version is number 7. Once you are happy with number 7, I think it's worth tidying up the file by removing some of these old functions. They will always live on in the version control system so deleting them from the current version of the file doesn't mean they will be gone forever!